examples/ensemble/plot_adaboost_multiclass.py

```diff
@@ -1,122 +1,235 @@
+depicted by Figure 1 in Zhu et al [1]_.
+
+The core principle of AdaBoost (Adaptive Boosting) is to fit a sequence of weak
+learners (e. g. Decision Trees) on repeatedly re-sampled versions of the data.
+Each sample carries a weight that is adjusted after each training step, such
+that misclassified samples will be assigned higher weights. The re-sampling with
+replacement then accounts for the different weights, because the chance of
+selecting a higher weighted sample several times for the new data set is bigger
+than a lower weighted sample to be selected. This way, each weak learner focuses
+on the difficult-to-classify samples during subsequent iterations.
+
+.. topic:: References:
+
+    .. [1] :doi:`J. Zhu, H. Zou, S. Rosset, T. Hastie, "Multi-class adaboost."
+           Statistics and its Interface 2.3 (2009): 349-360.
+           <10.4310/SII.2009.v2.n3.a8>`
 
 """
 
-# Author: Noel Dawe <noel.dawe@gmail.com>
-#
+# Noel Dawe <noel.dawe@gmail.com>
 # License: BSD 3 clause
 
-import matplotlib.pyplot as plt
-
+# %%
+# Creating the dataset
+# --------------------
+# The classification dataset is constructed by taking a ten-dimensional standard
+# normal distribution and defining three classes separated by nested concentric
+# ten-dimensional spheres such that roughly equal numbers of samples are in each
+# class (quantiles of the :math:`\chi^2` distribution).
 from sklearn.datasets import make_gaussian_quantiles
-from sklearn.ensemble import AdaBoostClassifier
-from sklearn.metrics import accuracy_score
-from sklearn.tree import DecisionTreeClassifier
 
 X, y = make_gaussian_quantiles(
-    n_samples=13000, n_features=10, n_classes=3, random_state=1
+    n_samples=2_000, n_features=10, n_classes=3, random_state=1
 )
 
-n_split = 3000
+# %%
+# We split the dataset into 2 sets: 70 percent of the samples are used for
+# training and the remaining 30 percent for testing.
+from sklearn.model_selection import train_test_split
 
-X_train, X_test = X[:n_split], X[n_split:]
-y_train, y_test = y[:n_split], y[n_split:]
-
-bdt_real = AdaBoostClassifier(
-    DecisionTreeClassifier(max_depth=2), n_estimators=300, learning_rate=1
+X_train, X_test, y_train, y_test = train_test_split(
+    X, y, train_size=0.7, random_state=42
 )
 
-bdt_discrete = AdaBoostClassifier(
-    DecisionTreeClassifier(max_depth=2),
-    n_estimators=300,
-    learning_rate=1.5,
+# %%
+# Training the `AdaBoostClassifier`
+# ---------------------------------
+# We train the :class:`~sklearn.ensemble.AdaBoostClassifier`. The estimator
+# utilizes boosting to improve the classification accuracy. Boosting is a method
+# designed to train weak learners (base estimators) that learn from their
+# predecessor's mistakes.
+#
+# Here, we define the weak learner as a
+# :class:`~sklearn.tree.DecisionTreeClassifier` and set the maximum number of
+# leaves to 8. In a real setting, this parameter should be tuned. We set it to a
+# rather low value to limit the runtime of the example.
```

Comment:
Isn't it that, the weak learners are mostly only stumps of a `DecisionTreeClassifier`? In which cases and how much can AdaBoost deviate from this and how big is the effect?

Reply:
A weak learner is a classifier that underfits. I think in the paper, it is defined as a learner that performs slightly better than the chance level. Decision stumps is an extreme case.

From my reading, if I recall correctly also, it is also expected that for the algorithm to behave properly (depending on the algorithm), the weak learner is required to have a minimum performance.

So in the end, I think that we can advise for searching the best hyperparameters. However, we can mitigate it by mentioning that AdaBoost is expected to use a weak learner and this should influence the definition of the parameter values to be searched.

Reply:
Yes, okay.